src/ceph-volume/ceph_volume/tests/test_terminal.py

```diff
@@ -66,3 +72,44 @@ def test_subcommand_found_and_dispatched_with_errors(self):
         with pytest.raises(SystemExit) as error:
             terminal.dispatch({'sub': BadSubCommand}, argv=['sub'])
         assert str(error.value) == '100'
+
+
+class TestWriteUnicode(object):
+
+    octpus_and_squid_en = u'octpus and squid'
+    octpus_and_squid_zh = u'章鱼和鱿鱼'
+    message = octpus_and_squid_en + octpus_and_squid_zh
+
+    def test_stdout_writer(self, capsys):
+        # should work with whatever stdout is
+        terminal.stdout(self.message)
+        out, _ = capsys.readouterr()
+        assert self.octpus_and_squid_en in out
+
+    def test_ascii_writer(self):
+        # should work with stdout with ascii encoding
+        self._test_writer('ascii')
+
+    def test_utf8_writer(self):
+        # should work with stdout with UTF-8 encoding
+        self._test_writer('utf8')
+
+    def _test_writer(self, encoding):
+        # mock a stdout with given encoding
+        buffer = io.BytesIO()
+        if sys.version_info >= (3, 0):
+            stdout = sys.stdout
+            stream = io.TextIOWrapper(buffer,
+                                      encoding=encoding,
+                                      errors=stdout.errors,
+                                      newline=stdout.newlines,
+                                      line_buffering=stdout.line_buffering)
+        else:
+            stream = codecs.getwriter(encoding)(buffer)
+            # StreamWriter does not have encoding attached to it, it will ask
+            # the inner buffer for "encoding" attribute in this case
+            stream.encoding = encoding
+        with mock.patch('sys.stdout', new=stream):
+            terminal.stdout(self.message)
+            sys.stdout.flush()
+            assert self.octpus_and_squid_en.encode(encoding) in buffer.getvalue()
```

Comment:
making this a fixture, and allowing buffer to be part of setup along with the message, this would become:


    @pytest.fixture
    def stream(buffer):
        # mock a stdout with given encoding
        if sys.version_info >= (3, 0):
            stdout = sys.stdout
            stream = io.TextIOWrapper(buffer,
                                      encoding=encoding,
                                      errors=stdout.errors,
                                      newline=stdout.newlines,
                                      line_buffering=stdout.line_buffering)
        else:
            stream = codecs.getwriter(encoding)(buffer)
            # StreamWriter does not have encoding attached to it, it will ask
            # the inner buffer for "encoding" attribute in this case
            stream.encoding = encoding
        return stream

    class TestWriteUnicode(object):

        def setup(self):
            octpus_and_squid_en = u'octpus and squid'
            octpus_and_squid_zh = u'章鱼和鱿鱼'
            self.message = octpus_and_squid_en + octpus_and_squid_zh
            self.buffer = io.BytesIO()

        def test_utf8_writer(self, stream, monkeypatch):
            stream = stream(self.buffer)
            monkeypatch.stattr('sys.stdout', stream)
            terminal.stdout(self.message)
            sys.stdout.flush()
            assert self.octpus_and_squid_en.encode('utf8') in self.buffer.getvalue()